cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java

```diff
@@ -383,22 +412,82 @@ public void apply() throws Exception {
     }
   }
 
+  // Returns true if the upgrade flag is set. Upgrade could have completed earlier than this since this flag is
+  // updated asynchronously.
+  public boolean isUpgradeComplete() {
+    return appVerUpgrade.get();
+  }
+
   /**
-   * Method to add version in JobKey and VersionKey in SchedulerStore.
+   * Method to add version to row key in SchedulerStore.
    *
-   * @throws Exception
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws DatasetManagementException
    */
-  public void upgrade() throws Exception {
-    // Create table instance.
-    initializeScheduleTable();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() throws Exception {
-          upgradeJobs();
-          upgradeTriggers();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
+    try {
+      factory.createExecutor(ImmutableList.of((TransactionAware) table))
+        .execute(new TransactionExecutor.Subroutine() {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
+              upgradeComplete.set(true);
+            }
+          }
+        });
+    } catch (TransactionFailureException e) {
+      // It is ok to ignore this error since this is an optimization check.
+      String errorMessage = "Initial check to see whether upgrade of TimeScheduleStore is " +
+        "complete or not, has failed. Proceeding with upgrade.";
+      if (e instanceof TransactionConflictException | e instanceof TransactionNotInProgressException) {
+        LOG.trace(errorMessage, e);
+      } else {
+        LOG.error(errorMessage, e);
+      }
+    }
+
+    // If upgrade is already complete, then simply return.
+    if (upgradeComplete.get()) {
+      return;
+    }
+
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of TimeScheduleStore.");
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
```

Comment:
We store the data in columns of a single row. And since by default the transaction conflict is set to ROW level, we can't do much. I will open a JIRA to migrate this data to rows since having 10k+ columns is not good.